### ISSUE TITLE
Unable to load "Fallout - Meshes.bsa" from New Vegas

### ISSUE THREAD

**Issue by Silverlan** [WrinklyNinja/libbsa#7]
All other .bsa-files from fallout new vegas can be loaded using _bsa_open_ just fine, however when trying to load the "Fallout - Meshes.bsa", it throws the exception from the following function:

        std::string BSA::getFileName(const uint8_t * fileNames, uint32_t offset) {
            const char * filename = reinterpret_cast<const char*>(fileNames + offset);

            //Find position of null character.
            char * nullTerminatorPos = strchr(const_cast<char*>(filename), '\0');
            if (nullTerminatorPos == NULL)
                throw error(LIBBSA_ERROR_PARSE_FAIL, "String at " + to_string(*(size_t*)filename) + "is not null terminated.");

            return ToUTF8(string(filename, nullTerminatorPos - filename));
        }

**Comment by WrinklyNinja** [comment]
The error is `Error message: "1w��" cannot be encoded in Windows-1252.` The quoted byte string is `0x7f 0x31 0x1f 0x77 0x8d 0xb3 0x02`, found at offset 0x47E4, in the middle of a data block, so libbsa is parsing the file incorrectly and expecting a file path in the wrong place.

The nearest preceding file path is `meshes\weapons\2handrifle` at 0x4596, and the following path is `meshes\architecture\nellis afb`.

**Comment by WrinklyNinja** [comment]
Here's a list of the folder and file names as parsed by libbsa. [filenames.txt](https://github.com/WrinklyNinja/libbsa/files/1056117/filenames.txt), up to the point at which this exception occurs. I tried comparing with `Fallout - Textures.bsa`, but libbsa also chokes on that.

If it's a simple fix, it's one that I'd need to read up on the BSA file format again to perform, which I'd rather not spend my free time doing. :grimacing: If you feel like fixing it yourself, you're welcome to, but otherwise there are plenty of other BSA parsers (some much better than this), as [Buddah](https://wrinklyninja.github.io/buddah/) can show you.

**Comment by Silverlan** [comment]
Actually, I noticed the error occurs randomly (With other .bsa-files as well) and I believe I found the cause.
Specifially, it's these two lines in "tes4bsa.cpp" (Lines 103 and 104):
`fileData.path += getFileName(fileNames, fileNameListPos);`
`fileNameListPos += fileData.path.length() + 1;`

_fileNameListPos_ should be incremented by the length of the file name, not the length of the entire path:
`auto fname = getFileName(fileNames, fileNameListPos);`
`fileData.path += fname;`
`fileNameListPos += fname.length() + 1;`

With that, _fileNameListPos_ ends up exactly at the same value as _header.totalFileNameLength_ after all file names have been read (Before it was incrementing past the array size), so that seems to fit.

I'm not familiar with the BSA format, so I'm not 100% sure about this, but I haven't had any problems since changing the code.
